### DISCUSSION TITLE
A Visual VSCT Designer extension

### DISCUSSION THREAD

**Discussion by yannduran** [VsixCommunity/Community.VisualStudio.Toolkit#21] [Ideas]
I got very excited when Mads mentioned the idea of a visual VSCT designer a couple of episodes ago because this is something that I've been very interested in since VS 2012.

Back in VS 2010 there was an extension that Microsoft published to the gallery called _VSPackage Designer_. I loved it and used to use it to design my VSCT files, which made the then very arcane VSCT file so much easier to deal with. There was almost no documentation on the VSCT back then.

I wanted so badly for it to be updated to VS 2012, but it was based on the _Microsoft Modelling SDK_, which in VS 2012 became the _DSL Tools_ (Domain-Specific Language). The _DSL Tools_ differed dramatically from the previous _Microsoft Modelling SDK_. So the extension wouldn't work without an extensive rewrite, and at that time nobody in Microsoft was interested in doing that. I had no experience with domain-specific languages, and back in those days we had no real direct access to anyone in Microsoft like we do today with Mads etc. Although I asked **several times** about the source code for extension, the answer that I got back was that nobody knew what had happened to it, as it had been written for Microsoft by an external company and was _lost_ (or so they thought).

When @justcla became a Program Manager on the VS Extensibility Team I mentioned this extension to him, and subsequently bugged him to try to find the source code for it. I wanted so badly to be the one who brought the extension back to life. Justin did eventually find the source code, and after bugging him some more, he managed to open source it internally. I then ended up doing some contract work for Justin shortly after that, and he agreed it was an important enough project for him to justify me working on it, and I was given access to the private open-sourced repo. I tried a number of things, including fork-lifting the definitions and custom code into a new VS 2015 DSL project, but sadly I just wasn't able to get it working in the time I had available to work on it. And I just didn't have the experience with the _DSL Tools_ to work on it after that either.

Although open-sourced internally, the repo is currently still private, but maybe @madskristensen could get it made public for us so it can be added to the _VSIX Community_ org. It could be called `Community.VisualStudio.Toolkit.Designer` (Visual Studio Community Toolkit Designer) or `Community.VisualStudio.VSCT.Designer` (Visual Studio VSCT Designer), and with the collective expertise of the _VSIX Community_ members, maybe there's someone out there who knows enough about domain-specific languages and using the _DSL Tools_ to help me to figure out why I couldn't get it to work.

**Comment by madskristensen** [comment]
This would be amazing. @justcla do you remember who owned the code?

**Reply by justcla** [reply]
I don't remember who the original owner/author was. But I think I found the code in a Microsoft repository. (Maybe VSTS at the time.) I probably copied it to GitHub under one of the Microsoft owned repos. I remember giving permission to Yann, as he was contracted to work on it.

**Comment by yannduran** [comment]
@madskristensen This is the **private** Microsoft repo that I forked when I was working on it.:

https://github.com/microsoft/VSPackageDesigner

I hope that helps you find it.

**Reply by reduckted** [reply]
That link 404s. Hopefully it's just a typo and not that it's been taken down. 🤞

**Reply by yannduran** [reply]
As I mentioned in my post, the repo is currently **_private_**. That's why you'd get a 404. I only meant for Mads to go looking for it, sorry. I've updated the message to hopefully make that clearer.

AFAIK I was the only person outside of Microsoft to be given access to it. Hopefully Mads will be able to change that so we can all have access to it.
